inferno-types/src/Inferno/Types/Type.hs

```diff
@@ -35,206 +35,33 @@ module Inferno.Types.Type
-          )
-          <+> "⇒"
-          <+> pretty ty
-
-newtype Subst = Subst (Map.Map TV InfernoType)
-  deriving stock (Eq, Ord)
-  deriving newtype (Semigroup, Monoid)
-
-instance Show Subst where
-  show (Subst m) = intercalate "\n" $ map (\(x, t) -> unpack $ renderPretty x <> " ~> " <> renderPretty t) $ Map.toList m
-
-class Substitutable a where
-  apply :: Subst -> a -> a
-  ftv :: a -> Set.Set TV
-
-instance Substitutable InfernoType where
-  apply _ (TBase a) = TBase a
-  apply (Subst s) t@(TVar a) = Map.findWithDefault t a s
-  apply s (t1 `TArr` t2) = apply s t1 `TArr` apply s t2
-  apply s (TArray t) = TArray $ apply s t
-  apply s (TSeries t) = TSeries $ apply s t
-  apply s (TOptional t) = TOptional $ apply s t
-  apply s (TTuple ts) = TTuple $ fmap (apply s) ts
-  apply s (TRep t) = TRep $ apply s t
-
-  ftv TBase {} = Set.empty
-  ftv (TVar a) = Set.singleton a
-  ftv (t1 `TArr` t2) = ftv t1 `Set.union` ftv t2
-  ftv (TArray t) = ftv t
-  ftv (TSeries t) = ftv t
-  ftv (TOptional t) = ftv t
-  ftv (TTuple ts) = foldr (Set.union . ftv) Set.empty ts
-  ftv (TRep t) = ftv t
-
-instance Substitutable ImplType where
-  apply s (ImplType impl t) =
-    ImplType (Map.map (apply s) impl) $ apply s t
-  ftv (ImplType impl t) = (foldr Set.union Set.empty $ map (ftv . snd) $ Map.toList impl) `Set.union` ftv t
-
-instance Substitutable TypeClass where
-  apply s (TypeClass n tys) = TypeClass n $ map (apply s) tys
-  ftv (TypeClass _ tys) = Set.unions $ map ftv tys
-
-instance Substitutable TCScheme where
-  apply (Subst s) (ForallTC as tcs t) = ForallTC as (Set.map (apply s') tcs) (apply s' t)
-    where
-      s' = Subst $ foldr Map.delete s as
-  ftv (ForallTC as tcs t) = ((ftv t) `Set.union` (Set.unions $ Set.elems $ Set.map ftv tcs)) `Set.difference` Set.fromList as
-
-instance Substitutable a => Substitutable [a] where
-  apply = map . apply
-  ftv = foldr (Set.union . ftv) Set.empty
-
-data Namespace
-  = FunNamespace Ident
-  | OpNamespace Ident
-  | EnumNamespace Ident
-  | ModuleNamespace ModuleName
-  | TypeNamespace Ident
-  deriving (Eq, Show, Ord, Generic, ToJSON, FromJSON)
-
-instance Pretty Namespace where
-  pretty = \case
-    FunNamespace (Ident i) -> pretty i
-    OpNamespace (Ident i) -> pretty i
-    EnumNamespace (Ident i) -> "#" <> pretty i
-    ModuleNamespace (ModuleName m) -> pretty m
-    TypeNamespace (Ident i) -> pretty i
-
-namespaceToIdent :: Namespace -> Ident
-namespaceToIdent = \case
-  FunNamespace i -> i
-  OpNamespace i -> i
-  EnumNamespace i -> i
-  TypeNamespace i -> i
-  ModuleNamespace _ -> error "namespaceToIdent undefined for ModuleNamespace"
-
-data TypeMetadata ty = TypeMetadata
-  { identExpr :: Expr () (),
-    docs :: Maybe Text,
-    ty :: ty
-  }
-  deriving (Eq, Show, Generic, Data, ToJSON, FromJSON)
+import Inferno.Types.Syntax
+  ( BaseType (..),
+    ImplType (..),
```

Comment:
I've left this module as is, in case there was some reason for separating these type definitions from the others in `Syntax.hs`. If there's no reason to keep this separate file, I can remove it.

Reply:
Syntax.hs is already pretty huge so might be good to keep some bits of it, like the types, separate

Reply:
Thanks, but since `Expr` now depends on `TCScheme`, there's going to be cyclic dependencies.. unless I try to reverse the dependency and have `Syntax.hs` import `Type.hs`? Some things like `TypeMetadata` might have to stay in `Syntax.hs` since it depends on `Expr`.

Reply:
ah good point... dunno may be worth making submodules for Expr.<...> for the auxiliary stuff like TypeMetadata...  I'll leave it up to you

Reply:
Okay thanks, I'll leave the restructuring for a future PR then. If no objections, I'll merge this PR in now?